### PR TITLE
Implement __cheri_offset and __cheri_addr casts

### PR DESCRIPTION
**Pull request by kgudka**
These casts allow you to obtain the offset and address, and follow the same style of `__cheri_ptr` (renamed from `__cheri_cast`) whereby you write the type after the keyword:

`long x = (__cheri_offset long)cap;`

and

`long y = (__cheri_addr long)cap;`

They get turned into calls to the intrinsics `llvm.cheri.cap.offset.get` and `llvm.cheri.cap.address.get` respectively.

`__cheri_cast` is still supported but results in a deprecated warning.

### CODE REVIEW

File: lib/CodeGen/CGExprScalar.cpp [hunk 1]
```diff
@@ -1810,6 +1810,33 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
...
+    // just the pointer to the referencee, which is what we do here.
+    if (const DeclRefExpr *DRE = dyn_cast<DeclRefExpr>(E)) {
+      if (const VarDecl *VD = dyn_cast<VarDecl>(DRE->getDecl())) {
+        if (const ReferenceType *RT = dyn_cast<ReferenceType>(VD->getType())) {
+					// RT is guaranteed to be a capability
+          Src = EmitLValue(E).getPointer();
+        }
+      }
+		}
+		if (!Src)
```
Comment (davidchisnall) [inline lib/CodeGen/CGExprScalar.cpp]: There's some weird indenting going on here.

File: lib/CodeGen/CGExprScalar.cpp [hunk 2]
```diff
@@ -1810,6 +1810,33 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
...
+    // emits a load of the referencee but that can lead to the Verifier being
+    // unhappy if the referencee type is incomplete. What we really want is
+    // just the pointer to the referencee, which is what we do here.
+    if (const DeclRefExpr *DRE = dyn_cast<DeclRefExpr>(E)) {
+      if (const VarDecl *VD = dyn_cast<VarDecl>(DRE->getDecl())) {
+        if (const ReferenceType *RT = dyn_cast<ReferenceType>(VD->getType())) {
+					// RT is guaranteed to be a capability
+          Src = EmitLValue(E).getPointer();
+        }
+      }
```
Comment (davidchisnall) [inline lib/CodeGen/CGExprScalar.cpp]: The outer `if` statements have only one line in the body, so LLVM style requires that they not have braces.

File: lib/CodeGen/CGExprScalar.cpp [hunk 3]
```diff
@@ -1810,6 +1810,33 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
...
+		}
+		if (!Src)
+      Src = Visit(E); // The default case
+    llvm::Type *ResultType = ConvertType(DestTy);
+    Src = Kind == CK_CHERICapabilityToOffset
+            ? CGF.getPointerOffset(Src)
+            : CGF.getPointerAddress(Src);
+    bool DestSigned = DestTy->isSignedIntegerOrEnumerationType();
+    // Insert int cast incase size of result type and capability offset field
```
Comment (davidchisnall) [inline lib/CodeGen/CGExprScalar.cpp]: incase is not a word.

File: lib/CodeGen/CGExprScalar.cpp [hunk 4]
```diff
@@ -1810,6 +1810,33 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
...
+    // are not the same. This will be a noop if the sizes are the same.
+    return Builder.CreateIntCast(Src, ResultType, DestSigned, "conv");
```
Comment (davidchisnall) [inline lib/CodeGen/CGExprScalar.cpp]: no-op or nop.

File: lib/Sema/SemaCast.cpp [hunk 5]
```diff
@@ -2945,20 +2945,80 @@ ExprResult Sema::BuildCheriCast(SourceLocation LParenLoc,
...
+                                           tok::TokenKind Kind, QualType
+                                           DestTy, TypeSourceInfo *TSInfo,
+                                           SourceLocation RParenLoc, Expr
+                                           *SubExpr) {
+  // Use getRealReferenceType() because getType() returns T for T&
+  QualType SrcTy = SubExpr->getRealReferenceType();
+  bool SrcIsCap = SrcTy->isCHERICapabilityType(Context);
+  if (!SrcIsCap) {
+    // XXXKG: What about functions?
+    Diag(SubExpr->getLocStart(), diag::err_cheri_offset_addr_invalid_source_type)
```
Comment (davidchisnall) [inline lib/Sema/SemaCast.cpp]: References to functions should always be pointer types in C (and LLVM IR).

File: lib/Sema/SemaCast.cpp [hunk 6]
```diff
@@ -2945,20 +2945,80 @@ ExprResult Sema::BuildCheriCast(SourceLocation LParenLoc,
...
+      << SrcTy << (Kind == tok::kw___cheri_offset);
+    return ExprError();
+  }
+
+  // XXXKG: Should pointer types also be allowed for __cheri_addr?
```
Comment (davidchisnall) [inline lib/Sema/SemaCast.cpp]: No, but a more-specific warning would be good (e.g. `__cheri_addr` is not valid for integral pointer types).

File: include/clang/Basic/TokenKinds.def [hunk 7]
```diff
@@ -571,8 +571,12 @@ KEYWORD(__bridge_retained            , KEYARC)
 KEYWORD(__bridge_retain              , KEYARC)
 
 
-// CHERI casts to and from capabilities
-KEYWORD(__cheri_cast                 , KEYALL)
+// CHERI casts.
+KEYWORD(__cheri_ptr                 , KEYALL) // cast ptr <-> capability
```
Comment (arichardson) [inline include/clang/Basic/TokenKinds.def]: I don't really like using `__cheri_ptr` to create capabilities: `(__cheri_ptr void* __capability)ptr;`?

How about adding `__cheri_cap` or maybe `__cheri_to_ptr`+`__cheri_to_cap`. Are there any cases where we need the current `__cheri_cast` sematics of converting either pointer -> cap or cap -> pointer? I don't think there were any in CheriBSD that did conversions both ways.

File: lib/Lex/PPMacroExpansion.cpp [hunk 8]
```diff
@@ -1268,7 +1268,7 @@ static bool HasFeature(const Preprocessor &PP, StringRef Feature) {
       .Case("underlying_type", LangOpts.CPlusPlus)
       .Case("capabilities", PP.getTargetInfo().SupportsCapabilities())
       .Case("pointer_interpretation", PP.getTargetInfo().SupportsCapabilities())
-      .Case("__cheri_cast", PP.getTargetInfo().SupportsCapabilities())
+      .Case("__cheri_ptr", PP.getTargetInfo().SupportsCapabilities())
```
Comment (arichardson) [inline lib/Lex/PPMacroExpansion.cpp]: The __has_feature check in CheriBSD will also need updating. Although we can probably just delete it now, I don't think CheriBSD compiles with any compiler version that doesn't include __cheri_cast

File: include/clang/Basic/TokenKinds.def (hunk 7 above)
Comment (kgudka) [reply include/clang/Basic/TokenKinds.def]: After talking with theraven, he suggested to add `__cheri_tocap` and `__cheri_fromcap`

**Comment by kgudka** [comment]
Should `__cheri_fromcap` and `__cheri_tocap` be allowed in purecap ABI mode? Would they just no-ops?

**Comment by davidchisnall** [comment]
Yes, because we can have a type declared with `#pragma pointer_interpretation integer` that we can cast to and from.  They're probably not sensible to use, because such pointers probably come from another DDC and so the result may not be sensible, but if the user does an explicit cast then we don't guarantee sanity...

**Comment by arichardson** [comment]
I think in the purecap ABI the pointer interpretation pragma changes the type to `long` in the AST so I don't think it will work. However, we do need them to be available as noops so that we can compile code both as hybrid and as purecap.

**Comment by davidchisnall** [comment]
I don't really like having them be nops.  You can `#define` them away in pure-cap mode if you know that's what you're doing, but having explicit casts that are silently ignored sounds like a good source of bugs.

**Comment by kgudka** [comment]
In the purecap ABI, they would only be no-ops if the source and destination types are capabilities and the types are compatible.

**Comment by arichardson** [comment]
Yes, in the purecap mode they can either be no-ops or errors due to incompatible types. I think this is better than just `#defining` it away to turn it into a C-style cast that can change the type.

**Comment by davidchisnall** [comment]
In the hybrid ABI, what happens if the source and destination are both capabilities and the types are compatible?  I really don't like this inconsistency - if you do an explicit cast and the destination type is the wrong kind of thing, then you're doing something wrong.  It's trivial for someone to put:

```C
#ifdef __CHERI_PURE_CAPABILITY__
#define __cheri_tocap
#define __cheri_fromcap
#endif
```
in a common header if they want this, but by making them silently nops we're letting the user write something that the compiler knows is wrong and silently accepting it.  At the very least, this should be a warning.

**Comment by kgudka** [comment]
I've currently implemented __cheri_fromcap and __cheri_tocap so that they would also be no-ops for hybrid code. We are currently outputting a warning for hybrid code but we can also output warnings for pure code as well

**Comment by arichardson** [comment]
This has been merged

**Pull request closed without merging**